### PR TITLE
Refactor archival workflow

### PR DESCRIPTION
**Pull request by yycptt**
- Refactor archiver workflow, handler and pump so that they can handle arbitrary archive requests.
- Use metrics scope instead of metrics client for archiver handler and pump
- Rename existing method and struct name to be more specific.
- Add replay workflow history test to ensure workflow backward compatibility.

### CODE REVIEW

File: common/metrics/defs.go [hunk 1]
```diff
@@ -1282,9 +1282,9 @@ var ScopeDefs = map[ServiceIdx]map[int]scopeDefinition{
 		IndexProcessorScope:                 {operation: "IndexProcessor"},
 		ArchiverDeleteHistoryActivityScope:  {operation: "ArchiverDeleteHistoryActivity"},
 		ArchiverUploadHistoryActivityScope:  {operation: "ArchiverUploadHistoryActivity"},
-		ArchiverScope:                       {operation: "Archiver"},
-		ArchiverPumpScope:                   {operation: "ArchiverPump"},
-		ArchiverArchivalWorkflowScope:       {operation: "ArchiverArchivalWorkflow"},
+		HistoryArchivalHandlerScope:         {operation: "HistoryArchivalHandler"},
```
Comment (andrewjdawson2016) [inline common/metrics/defs.go]: Can you create ticket or just a note to update the dashboard.

File: service/worker/archiver/pump.go [hunk 2]
```diff
@@ -37,42 +37,45 @@ type (
 	// PumpResult is the result of pumping requests into request channel
 	PumpResult struct {
 		PumpedHashes          []uint64
-		UnhandledCarryover    []ArchiveRequest
+		UnhandledCarryover    []interface{}
```
Comment (andrewjdawson2016) [inline service/worker/archiver/pump.go]: Is this backwards compatible?

File: service/worker/archiver/request_processor.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+// NewHistoryRequestProcessor returns a new processor for history requests
+func NewHistoryRequestProcessor(logger log.Logger, metricsScope metrics.Scope) RequestProcessor {
+	return &historyRequestProcessor{
+		logger:       logger,
+		metricsScope: metricsScope,
+	}
+}
+
+func (p *historyRequestProcessor) Process(ctx workflow.Context, request interface{}) {
```
Comment (andrewjdawson2016) [inline service/worker/archiver/request_processor.go]: Is this code identical to what was in handler.go before?

File: service/worker/archiver/request_receiver.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+	historyRequestReceiver struct{}
+)
+
+// NewHistoryRequestReceiver returns a new receiver for archive history requests
+func NewHistoryRequestReceiver() RequestReceiver {
+	return &historyRequestReceiver{}
+}
+
+func (r *historyRequestReceiver) Receive(ctx workflow.Context, ch workflow.Channel) (interface{}, bool) {
```
Comment (andrewjdawson2016) [inline service/worker/archiver/request_receiver.go]: Should channel be a field?

File: service/worker/archiver/workflow_test.go [hunk 5]
```diff
@@ -104,34 +118,58 @@ func (s *workflowSuite) TestArchivalWorkflow_Exit_TimeoutWithoutSignals() {
...
+func (s *workflowSuite) TestReplayArchiveHistoryWorkflow() {
+	logger, _ := zap.NewDevelopment()
+	globalLogger = workflowTestLogger
+	globalMetricsClient = metrics.NewClient(tally.NewTestScope("replay", nil), metrics.Worker)
+	globalConfig = &Config{
+		ArchiverConcurrency:           dynamicconfig.GetIntPropertyFn(50),
+		ArchivalsPerIteration:         dynamicconfig.GetIntPropertyFn(1000),
+		TimeLimitPerArchivalIteration: dynamicconfig.GetDurationPropertyFn(MaxArchivalIterationTimeout()),
+	}
+	err := worker.ReplayWorkflowHistoryFromJSONFile(logger, "testdata/archive_history_workflow_history.json")
```
Comment (andrewjdawson2016) [inline service/worker/archiver/workflow_test.go]: Can you chat with Maxim about how these files should be named? It seems like each time we make a workflow change we want to add a history to test and those histories need some nice name?

File: service/worker/archiver/pump.go (hunk 2 above)
Comment (yycptt) [reply service/worker/archiver/pump.go]: Tested locally. It's backward compatible.

**Pull request merged**